File: apps/platform/app/Platform/Support/Money/MinorAmount.php
<?php

declare(strict_types=1);

namespace App\Platform\Support\Money;

use Brick\Math\RoundingMode;
use Brick\Money\Money;

final class MinorAmount
{
    private const BASIS_POINTS_SCALE = 10000;

    public static function calculateTax(
        int $netMinor,
        int $rateBasisPoints,
        bool $inclusive,
        string $currency,
    ): int {
        if ($rateBasisPoints <= 0 || $netMinor <= 0) {
            return 0;
        }

        $amount = self::ofMinor($netMinor, $currency);

        if ($inclusive) {
            $exclusiveBase = $amount
                ->multipliedBy(self::BASIS_POINTS_SCALE)
                ->dividedBy(self::BASIS_POINTS_SCALE + $rateBasisPoints, RoundingMode::HALF_UP);

            return self::toMinor($amount->minus($exclusiveBase));
        }

        return self::toMinor(
            $amount
                ->multipliedBy($rateBasisPoints)
                ->dividedBy(self::BASIS_POINTS_SCALE, RoundingMode::HALF_UP),
        );
    }

    public static function percentageOf(
        int $amountMinor,
        int $basisPoints,
        string $currency,
        RoundingMode $roundingMode = RoundingMode::DOWN,
    ): int {
        if ($amountMinor <= 0 || $basisPoints <= 0) {
            return 0;
        }

        return self::toMinor(
            self::ofMinor($amountMinor, $currency)
                ->multipliedBy($basisPoints)
                ->dividedBy(self::BASIS_POINTS_SCALE, $roundingMode),
        );
    }

    /**
     * @param  array<int, int>  $baseAmounts
     * @return array<int, int>
     */
    public static function allocateAcross(array $baseAmounts, int $allocationMinor): array
    {
        if ($allocationMinor <= 0 || $baseAmounts === []) {
            return array_fill(0, count($baseAmounts), 0);
        }

        $baseTotalMinor = array_sum($baseAmounts);

        if ($baseTotalMinor <= 0) {
            return array_fill(0, count($baseAmounts), 0);
        }

        $allocations = [];
        $allocatedMinor = 0;
        $lastIndex = array_key_last($baseAmounts);

        foreach ($baseAmounts as $index => $baseAmount) {
            if ($index === $lastIndex) {
                $allocations[$index] = max(0, min($allocationMinor - $allocatedMinor, $baseAmount));

                continue;
            }

            $allocation = intdiv($allocationMinor * $baseAmount, $baseTotalMinor);
            $allocations[$index] = max(0, min($allocation, $baseAmount));
            $allocatedMinor += $allocations[$index];
        }

        return array_values($allocations);
    }

    public static function ratioThreshold(int $amountMinor, int $basisPoints, string $currency): int
    {
        return self::percentageOf($amountMinor, $basisPoints, $currency, RoundingMode::HALF_UP);
    }

    private static function ofMinor(int $amountMinor, string $currency): Money
    {
        return Money::ofMinor($amountMinor, $currency);
    }

    private static function toMinor(Money $money): int
    {
        return $money->getMinorAmount()->toInt();
    }
}

File: apps/pos/lib/src/core/support/minor_amount.dart
final class MinorAmount {
  static const int basisPointsScale = 10000;

  static int calculateTax({
    required int netMinor,
    required int rateBasisPoints,
    required bool inclusive,
  }) {
    if (netMinor <= 0 || rateBasisPoints <= 0) {
      return 0;
    }

    if (inclusive) {
      final exclusiveBase = _divideAndRoundHalfUp(
        dividend: netMinor * basisPointsScale,
        divisor: basisPointsScale + rateBasisPoints,
      );

      return netMinor - exclusiveBase;
    }

    return _divideAndRoundHalfUp(
      dividend: netMinor * rateBasisPoints,
      divisor: basisPointsScale,
    );
  }

  static int percentageOf({
    required int amountMinor,
    required int basisPoints,
  }) {
    if (amountMinor <= 0 || basisPoints <= 0) {
      return 0;
    }

    return (amountMinor * basisPoints) ~/ basisPointsScale;
  }

  static List<int> allocateAcross(List<int> baseAmounts, int allocationMinor) {
    if (allocationMinor <= 0 || baseAmounts.isEmpty) {
      return List<int>.filled(baseAmounts.length, 0, growable: false);
    }

    final baseTotalMinor = baseAmounts.fold<int>(0, (sum, item) => sum + item);

    if (baseTotalMinor <= 0) {
      return List<int>.filled(baseAmounts.length, 0, growable: false);
    }

    final allocations = List<int>.filled(
      baseAmounts.length,
      0,
      growable: false,
    );
    var allocatedMinor = 0;

    for (var index = 0; index < baseAmounts.length; index += 1) {
      if (index == baseAmounts.length - 1) {
        allocations[index] = (allocationMinor - allocatedMinor).clamp(
          0,
          baseAmounts[index],
        );
        continue;
      }

      final allocation =
          (allocationMinor * baseAmounts[index]) ~/ baseTotalMinor;
      allocations[index] = allocation.clamp(0, baseAmounts[index]);
      allocatedMinor += allocations[index];
    }

    return allocations;
  }

  static int _divideAndRoundHalfUp({
    required int dividend,
    required int divisor,
  }) {
    if (divisor == 0) {
      throw ArgumentError.value(divisor, 'divisor', 'Must not be zero.');
    }

    final sign = dividend.sign * divisor.sign;
    final absoluteDividend = dividend.abs();
    final absoluteDivisor = divisor.abs();
    var quotient = absoluteDividend ~/ absoluteDivisor;
    final remainder = absoluteDividend % absoluteDivisor;

    if (remainder * 2 >= absoluteDivisor) {
      quotient += 1;
    }

    return sign < 0 ? -quotient : quotient;
  }
}
